### ISSUE TITLE
Operation too slow. Less than 1000 bytes/sec transferred the last 30 seconds

### ISSUE THREAD

**Issue by brianjmurrell** [open-sw/pkg-cacher#39]
When using `pkg-cacher` the clients intermittently experience the dreaded:

```
[MIRROR] libreoffice-core-6.0.6.2-2.fc28.x86_64.rpm: Curl error (28): Timeout was reached for http://pkgs.example.com:3142/fedora/updates/28/Everything/x86_64/Packages/l/libreoffice-core-6.0.6.2-2.fc28.x86_64.rpm [Operation too slow. Less than 1000 bytes/sec transferred the last 30 seconds]
[FAILED] libreoffice-core-6.0.6.2-2.fc28.x86_64.rpm: No more mirrors to try - All mirrors were already tried without success                                                           
```

I suspect that this is because for large[r] packages, it takes more than 30 seconds for pkg-cacher to fetch the package and cache it before serving it back out to the client that requested it.

Probably the ideal solution is to start writing what is read during the fetch from the upstream package server to the client simultaneously with writing it to the local cache.  Hopefully that makes sense.  Maybe to clarify I could add that this would mean switching from the serial steps of reading from upstream, writing to the local cache, reading from the local cache, writing to the client to parallelizing the writes to the local cache with writing to the client, so that each block read from the upstream server is actually written twice in the same loop, once to the local cache file and once to the client socket.

**Comment by ganto** [comment]
To avoid this, I simply raised the timeout on the clients. Since then I never had an issue anymore.

**Comment by brianjmurrell** [comment]
Sure, but that's a hack at best and causes added delay when there really is a problem.

This is something that the proxy should deal with properly.

**Comment by ganto** [comment]
I totally agree with you. That's why I opened #38. I'm still using pkg-cacher on a daily basis and would love if there would be some activity in the upstream repository again.
